crates/shared/src/trade_finding.rs

```diff
@@ -49,6 +120,114 @@ pub struct Trade {
+            .context("clearing sell price missing")?
+            .to_big_rational();
+        let buy_price = self
+            .clearing_prices
+            .get(buy_token)
+            .context("clearing buy price missing")?
+            .to_big_rational();
+        let order_amount = in_amount.to_big_rational();
+
+        let out_amount = match order_kind {
+            OrderKind::Sell => order_amount
+                .mul(&sell_price)
+                .checked_div(&buy_price)
+                .context("div by zero: buy price")?,
+            OrderKind::Buy => order_amount
+                .mul(&buy_price)
+                .checked_div(&sell_price)
+                .context("div by zero: sell price")?,
+        };
+
+        big_rational_to_u256(&out_amount).context("out amount is not a valid U256")
+    }
+
+    /// Calculate the net token changes for all the JIT orders.
+    /// Positive values mean the solver gains the token, negative values mean
+    /// the solver loses it.
+    pub fn jit_orders_net_token_changes(&self) -> Result<HashMap<H160, BigRational>> {
+        let mut net_token_changes: HashMap<H160, BigRational> = HashMap::new();
+
+        for jit_order in self.jit_orders.iter() {
+            let sell_price = *self
+                .clearing_prices
+                .get(&jit_order.sell_token)
+                .context("JIT order sell token clearing price is missing")?;
+            let buy_price = *self
+                .clearing_prices
+                .get(&jit_order.buy_token)
+                .context("JIT order buy token clearing price is missing")?;
+            let executed_amount = jit_order.executed_amount;
+
+            let (executed_sell, executed_buy) = match jit_order.side {
+                Side::Sell => {
+                    let buy_amount = executed_amount
+                        .checked_mul(sell_price)
+                        .context("multiplication overflow in JIT order sell")?
+                        .checked_ceil_div(&buy_price)
+                        .context("division by zero in JIT order sell")?;
+                    (
+                        executed_amount.to_big_rational(),
+                        buy_amount.to_big_rational(),
+                    )
+                }
+                Side::Buy => {
+                    let sell_amount = executed_amount
+                        .checked_mul(buy_price)
+                        .context("multiplication overflow in JIT order buy")?
+                        .checked_ceil_div(&sell_price)
+                        .context("division by zero in JIT order buy")?;
+                    (
+                        sell_amount.to_big_rational(),
+                        executed_amount.to_big_rational(),
+                    )
+                }
+            };
+
+            net_token_changes
+                .entry(jit_order.sell_token)
+                .and_modify(|e| *e += &executed_sell)
+                .or_insert(executed_sell);
```

Comment:
nit: I think using `.or_default()` and always doing the modification seems a bit easier to understand. Also this allows to move the `.to_big_rational()` here so that the tuples returned from the `match` can be 1 line.
```suggestion
            net_token_changes
                .entry(jit_order.sell_token)
                .or_default()
                .add_assign(executed_sell.to_big_rational());
```